[OpenMP] Enable automatic unified shared memory on MI300A.

carlobertolli opened the pull request:
This patch enables applications that did not request OpenMP unified_shared_memory to run with the same zero-copy behavior, where mapped memory does not result in extra memory allocations and memory copies, but CPU-allocated memory is accessed from the device. The name for this behavior is "automatic zero-copy" and it relies on detecting: that the runtime is running on a MI300A, that the user did not select unified_shared_memory in their program, and that XNACK (unified memory support) is enabled in the current GPU configuration. If all these conditions are met, then automatic zero-copy is triggered.

This patch also introduces an environment variable OMPX_APU_MAPS that, if set, triggers automatic zero-copy also on non APU GPUs (e.g., on discrete GPUs).
This patch is still missing support for global variables, which will be provided in a subsequent patch.

carlobertolli committed:
[OpenMP] Enable automatic unified shared memory on MI300A.

This patch enables applications that did not request OpenMP unified_shared_memory to run with the same zero-copy behavior,
where mapped memory does not result in extra memory allocations and memory copies, but CPU-allocated memory is accessed from the device.
The name for this behavior is "automatic zero-copy" and it relies on detecting: that the runtime is running on a MI300A,
that the user did not select unified_shared_memory in their program, and that XNACK (unified memory support) is enabled in the current GPU configuration.
If all these conditions are met, then automatic zero-copy is triggered.

This patch also introduces an environment variable OMPX_APU_MAPS that, if set, triggers automatic zero-copy
also on non APU GPUs (e.g., on discrete GPUs).
This patch is still missing support for global variables, which will be provided in a subsequent patch.

Co-authored-by: Thorsten Blass <thorsten.blass@amd.com>
Co-authored-by: Carlo Bertolli <carlo.bertolli@amd.com>

carlobertolli commented:
Reland #77512

carlobertolli commented:
Updated comment to describe when automatic zero-copy is triggered for AMD gpus.

arsenm commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -2728,6 +2749,30 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
+    std::string StrGfxName(ComputeUnitKind);
+    std::transform(std::begin(StrGfxName), std::end(StrGfxName),
+                   std::begin(StrGfxName),
+                   [](char c) { return std::tolower(c); });
```
I would hope only lowercased names would reach here to begin with

carlobertolli replied:
Do you mean that we don't need to lowercase the names because they already are?

jhuber6 replied:
Yeah I doubt we need to sanitize this.

arsenm commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -2728,6 +2749,30 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
 
+  /// Detect if current architecture is an APU.
+  Error checkIfAPU() {
+    std::string StrGfxName(ComputeUnitKind);
```
Shouldn't really need std::string for a small set of constant strings?

carlobertolli replied:
Should I use
const std::string StrGfxName(ComputeUnitKind);
?
Or how do you propose I change the code?

arsenm replied:
StringRef?

arsenm commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -2728,6 +2749,30 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
+                   std::begin(StrGfxName),
+                   [](char c) { return std::tolower(c); });
+    if (StrGfxName == "gfx940") {
+      IsAPU = true;
```
Can we avoid depending on listing device names?

carlobertolli replied:
What other ways do we have to detect an APU? It is either a gfx940 or a gfx942 with a specific chip ID. I don't see any other way to determine this.

jhuber6 replied:
Really wish HSA exported some stable ID to use rather than strings, like the ones in the ELF headers.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -1848,6 +1848,7 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
         OMPX_StreamBusyWait("LIBOMPTARGET_AMDGPU_STREAM_BUSYWAIT", 2000000),
         OMPX_UseMultipleSdmaEngines(
             "LIBOMPTARGET_AMDGPU_USE_MULTIPLE_SDMA_ENGINES", false),
+        HSAXnackEnv("HSA_XNACK", false), OMPX_ApuMaps("OMPX_APU_MAPS", false),
```
I really don't think we should be parsing HSA's environment variables redundantly. This information is available from the system target in the form of `amdgcn-amd-amdhsa-gfx...:xnack+` right? We should use that instead.

carlobertolli replied:
Which ROCr function returns that string for the system (and not for the binary)?

jhuber6 replied:
hsa_isa_get_info_alt with HSA_ISA_INFO_NAME. I went through this code recently w/ Saiyedul and it should respect the environment variable AFAIK.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/dynamic_hsa/hsa_ext_amd.h:
```diff
@@ -63,6 +63,7 @@ typedef enum {
 } hsa_amd_memory_pool_access_t;
 
 typedef enum hsa_amd_agent_info_s {
+  HSA_AMD_AGENT_INFO_CHIP_ID = 0xA000,
```
Just checking, is this a new thing? I remember having issues when something came from a really new HSA library, I think from the SDMA patch.

carlobertolli replied:
It's not, I believe. The low value should be indicative.

jhuber6 commented on openmp/libomptarget/include/Shared/Requirements.h:
```diff
@@ -33,7 +33,12 @@ enum OpenMPOffloadingRequiresDirFlags : int64_t {
+  /// when running on an APU, the GPU plugin may decide to
+  /// run in zero-copy even though the user did not program
+  /// their application with unified_shared_memory requirement.
+  OMPX_REQ_AUTO_ZERO_COPY = 0x020
```
Is this the right place for this? These enumerations are supposed to correspond directly to `omp requires xxx` directives.

carlobertolli replied:
If you look at the previous review #75999 there was consensus of having this here. I understand this does not match 1-to-1 what OpenMP requires clauses have, but it also means that we do not have to store it in some boolean nor to keep asking the plugin about it.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -2728,6 +2749,30 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
+    std::transform(std::begin(StrGfxName), std::end(StrGfxName),
+                   std::begin(StrGfxName),
+                   [](char c) { return std::tolower(c); });
+    if (StrGfxName == "gfx940") {
```
This is pretty gross, I really wish this was exposed in a non-string way if it's not. If there's no way to make this any better just use a `StringSwitch`.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -2794,6 +2846,9 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
   uint64_t StackSize = 16 * 1024 /* 16 KB */;
+
+  /// Is the plugin associated with an APU?
+  bool IsAPU{false};
```
Weird initializer, just use equals.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/utils/UtilitiesRTL.h:
```diff
@@ -116,6 +116,34 @@ inline bool isImageCompatibleWithEnv(StringRef ImageArch, uint32_t ImageFlags,
   return true;
 }
 
+inline bool isXnackEnabledViaKernelParam() {
```
Could use a comment explaining this weird file magic and what its purpose is.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/utils/UtilitiesRTL.h:
```diff
@@ -116,6 +116,34 @@ inline bool isImageCompatibleWithEnv(StringRef ImageArch, uint32_t ImageFlags,
+  StringRef RefString("amdgpu.noretry=");
+  int SizeOfRefString = RefString.size();
+
+  size_t Pos = FileContent.find_insensitive(RefString);
```
StringRef has `contains_insensitive`.

carlobertolli commented:
I believe I have addressed all concerns with the latest update. I also added a small refactoring of common code in the amdgpu plugin, used to obtain a string with target triple and features.
Thanks for the comments!

jhuber6 commented on openmp/libomptarget/include/Shared/PluginAPI.h:
```diff
@@ -219,6 +219,9 @@ int32_t __tgt_rtl_initialize_record_replay(int32_t DeviceId, int64_t MemorySize,
                                            uint64_t &ReqPtrArgOffset);
+
+// Returns true if the device \p DeviceId suggests to use auto zero-copy.
+int32_t __tgt_rtl_use_auto_zero_copy(int32_t DeviceId);
```
Unrelated, but we really need a common method to get target specific information. Something I'm planning on working on when we try to do the whole offload thing.

carlobertolli replied:
agreed.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -183,6 +183,29 @@ Error asyncMemCopy(bool UseMultipleSdmaEngines, void *Dst, hsa_agent_t DstAgent,
 #endif
 }
 
+Expected<StringRef> getTargetTripleAndFeatures(hsa_agent_t Agent) {
```
Don't think this can be a `StringRef`, as it's referencing the string in `llvm::SmallVector<char> ISAName(Length);` which goes out of scope.

carlobertolli replied:
Right, I changed that to std::string, thanks for the catch!

jhuber6 commented on openmp/libomptarget/src/PluginManager.cpp:
```diff
@@ -144,19 +144,32 @@ void PluginAdaptorTy::initDevices(PluginManager &PM) {
   ExclusiveDevicesAccessor->reserve(DeviceOffset + NumPD);
+  // Auto zero-copy is a per-device property. We need to ensure
+  // that all devices are suggesting to use it.
+  bool UseAutoZeroCopy = true;
+  if (NumPD == 0)
+    UseAutoZeroCopy = false;
```
nit. could initialize this to `NumPD == 0`.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -3039,30 +3130,15 @@ struct AMDGPUPluginTy final : public GenericPluginTy {
+          utils::getTargetTripleAndFeatures(Agent);
+      if (!TargeTripleAndFeaturesOrError)
+        return TargeTripleAndFeaturesOrError.takeError();
+      StringRef TargeTripleAndFeatures(*TargeTripleAndFeaturesOrError);
```
Shouldn't need this intermediate, std::string implicitly converts to StringRef.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -2728,6 +2780,34 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
   }
 
+  /// Detect if current architecture is an APU.
+  Error checkIfAPU() {
```
I was talking with SuiChi about checking this. He says that HSA_PROFILE_FULL should be set if the pointters are usable https://alt-doc-doc-test.readthedocs.io/en/latest/ROCm_API_References/ROCr-Runtime.html#group__agentinfo_1gacafd4247e2a04cbe0ac0b3998c127532. This is apparently how it's used in rocm-info https://github.com/ROCm/ROCR-Runtime/blob/master/src/core/runtime/amd_gpu_agent.cpp#L115. So you need to query agent info with the profile as an argument.

carlobertolli replied:
Host-allocated ppointers are usable on any gpu that supports unified memory, including MI200 and others.
This is not what we are determining here. What we are finding is whether the architecture is an APU, not a discrete GPU with unified memory capability. If this is changed to return true on MI200, it will make automatic zero-copy the default behavior on that gpu if XNACK is enabled. That's not what we want. What we want is to enable automatic zero-copy on APUs by default, but not enable it by default on discrete GPUs.
This logic follows an observation that OpenMP applications on discrete GPUs are *typically* optimized for data movement, and that configuration performs better than driver-level paging.
HIP doesn't have a concept of automatic zero-copy, as far as I know.

jhuber6 replied:
Well maybe this function should be renamed then? Because I saw that the linked code above used that method to determine `is_apu_node`, which seemed identical to what this is trying to determine. So this profile is set based off of `is_apu_node` in the above source. So is this referring to something different?

carlobertolli replied:
The patch (or libomptarget after patching it) doesn't have a "is_apu_node" name in it (either function or variable). I wonder if you are getting confused by successive amend's or if you are referring to downstream?

jhuber6 replied:
I'm just confused, because the name suggests that the full profile implies we have USM, which is what's necessary for this behavior. The ROCR code in question uses the same logic to set a variable called `isAPU` which suggests the same logic t his `checkIfAPU` is trying to accomplish. I guess the problem is that this is more general than what we're going for here and merely want it for this single MI300 architecture?

carlobertolli replied:
Leaving a comment here about investigations into rocr behavior: that code does not apply to MI300A. A new API to be able to detect an APU might be available soon, but in the meantime we need to go with the code that we have. I will modify to use the relevant ROCr API if and when it becomes available.

jhuber6 replied:
Alright, maybe put a TODO here to update this function once said feature is available or something.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -1940,6 +1964,19 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
+        utils::getTargetTripleAndFeatures(Agent);
+    if (!TargeTripleAndFeaturesOrError)
+      return TargeTripleAndFeaturesOrError.takeError();
+    StringRef TargeTripleAndFeatures(*TargeTripleAndFeaturesOrError);
+    if (TargeTripleAndFeatures.contains("xnack+"))
```
Nit, it's just do `->contains` and save the extra var.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -183,6 +183,29 @@ Error asyncMemCopy(bool UseMultipleSdmaEngines, void *Dst, hsa_agent_t DstAgent,
+    llvm::StringRef TripleTarget(ISAName.begin(), Length);
+    if (TripleTarget.consume_front("amdgcn-amd-amdhsa"))
+      Target = TripleTarget.ltrim('-').rtrim('\0').str();
+    return HSA_STATUS_SUCCESS;
```
I never looked at this much, what's the expected result here if there are multiple ISAs? Isn't there `HSA_STATUS_INFO_BREAK` to make the iterator stop?

carlobertolli replied:
We really loop over a single agent and I checked with rocr team: it can only resolve in a single isa. I am guessing that the HSA API (of which ROCr is only an implementation) leaves room for multiple ISAs associated with each agent, but that doesn't happen in this case.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -2728,6 +2780,34 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
+                        .Default(false);
+    if (!MayBeAPU)
+      return Plugin::success();
+    else {
```
What's with this `else`? If we returned then it shouldn't be necessary.

jhuber6 reviewed (commented):
A few final nits and I'll approve it.

jhuber6 reviewed (approved):
LG, Thanks
